internal/chain/engine.go

```diff
@@ -327,16 +327,41 @@ func (e *engineImpl) Finalize(
 		chain, state, header, e.Beaconchain(), sigsReady,
 	)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Apply slashes
 	if isBeaconChain && inStakingEra && len(doubleSigners) > 0 {
 		if err := applySlashes(chain, header, state, doubleSigners); err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
 	} else if len(doubleSigners) > 0 {
-		return nil, nil, errors.New("slashes proposed in non-beacon chain or non-staking epoch")
+		return nil, nil, nil, errors.New("slashes proposed in non-beacon chain or non-staking epoch")
+	}
+
+	// must be done after slashing and paying out rewards
+	// to avoid conflicts with snapshots
+	// any nil (un)delegations are eligible to be removed
+	// bulk pruning happens at the end of the NoNilDelegationsEpoch
+	// and then continuous pruning happens at the end of every epoch thereafter
```

Comment:
Do we have a separate continuous pruning besides bulk pruning? It's the same logic that happens at the end of epochs right?

Reply:
The same logic is applied at the end of each epoch. Regular / continuous pruning can occur only after undelegations are paid out; therefore it must occur at the end of the epoch.

Reply:
For a machine which inserts blocks at 200-600 ms, for an average of 120 blocks per minute:
- Bulk pruning took 41ms for the validator wrappers on chain. The number of validator wrappers thus modified were 490.
- Updating the delegation indices (off chain data) took 1061ms. The number of delegators impacted was 49000.
Total extra time at hard fork block is 1.1s. If we keep only bulk pruning, the time at each epoch block after the hard fork block is bound to be lower, and we can have code that is less complicated to maintain.
Total extra time spent per epoch is 1.1s, or lower.

On the same machine, continuous pruning results are as per below.
- Continuous pruning requires the modification of validators 1000 times per 2800 blocks (spread across multiple transactions), and thus per epoch approximately ~12,000 validators. This modification takes an additional time of roughly 960ms per epoch.
- This continuous pruning impacted 26,000 delegators across 2800 blocks. Over an epoch this is 304,000 changes. The time taken then, per epoch, is additional 6.8s.
Total extra time per epoch is thus 7.8s.

Based on the above information, it makes sense to go for solely the bulk pruning approach at the end of each epoch. I will push that out after I am done with the two precompiles.